### ISSUE TITLE
Channel lifetime management

### ISSUE THREAD

**Issue by akoshelev** [private-attribution/ipa#168]
Currently, we have the ability to open a communication channel per step between two helper parties, but we lack a mechanism to close it. Once channel is open, it stays as such for the duration of the query on both sending and receiving side, consuming resources and CPU time if we poll them.

This problem is not trivial to solve because the decision to close it should be made on the protocol side (only that code knows when the channel is no longer in use) but this decision must be propagated through all the layers on both sending and receiving side, so both can actually destroy all the resources they hold to keep the channel open. 

It may be the same solution as for `prss` to avoid re-using the same index more than once or we could invent another mechanism to do that. Drop guards sound really nice if we can make it work. Channels will be closed automatically after protocols terminate but likely that is not the only possible solution here.

**Comment by akoshelev** [comment]
Had a chat with @martinthomson about it. This seems closely related to another issue we've discovered where setting batch size to a value greater than 10 and input size >> 10 led to some protocols being stuck. For example, malicious validator only needs to send two records as part of validate: https://github.com/private-attribution/ipa/blob/00130133092e0368eb0b35a2230915d815c7578a/src/protocol/malicious.rs#L202

These related problems therefore must be addressed by any solution we propose here. 

## Solutions

The idea is to (surprise) close the channel when it is no longer needed. As only protocols know whether the channel is required or no, this decision must be made on the application layer. This solves the batching problem as well because infra does not need to be kept in the dark whether it should receive more messages or no - it can simply flush everything it has down to the network.

### [Broken] Explicit channel closure

Just to illustrate the problem, lets briefly see why `channel.close()` is a footgun in our model


```rust
       #[tokio::test]
    pub async fn broken_closure() {
        use futures::StreamExt;

        let world = TestWorld::new(QueryId);
        let c = world.contexts::<Fp31>();
        let c0 = &c[0];
        // channel is shared and can be requested at any time.
        let channel1 = c0.mesh(); 
        let channel2 = c0.mesh();
        let mut futures = FuturesOrdered::new();

        for i in 0..10 {
            let f = channel1.send(Role::H2, RecordId::from(i), Fp31::from(i as u128));
            futures.push_back(f);
        }

        channel2.close(); // it explodes here
        let results = futures.collect::<Vec<Result<(), crate::helpers::Error>>>();
    }

```

Rust borrow checker cannot help us because `channel2` points to the same channel, but has no real reference to it. Closing it is fine from Rust's perspective, but it will lead to messages being dropped on the floor. 


### Channel size hint 

One way to do it is to hint Infra how many records in total it should expect

```rust
    let channel = ctx.mesh(1000); // after record_id =. 999 is received, the channel will be automatically closed
```

the downside is that all protocols that request access to this channel need to provide this value

```rust
   let channel = ctx.narrow(&MyStep::DoSomething).mesh(1000); // opens up a new channel with total capacity for 1000 records
   // ....
   // somewhere else deep inside the call stack
   let channel = ctx.narrow(&MyStep::DoSomething).mesh(2000); // ignored? rejected? resized?
```

### RAII guard 

much more ergonomic way to deal with it is to introduce another method that returns an owned handle of the channel. That allows to use RAII guard to close it


```rust
     {
          let channel = ctx.narrow(&MyStep::DoSomething).mesh_owned().await.unwrap(); // Infra needs to make sure there is only one owned channel handle
          let shared_channel = ctx.
          // use channel. it can also be moved by value or borrowed
     } // channel is closed and flushed here. 
```

Infra can be smart and prevent this channel from opening again by keeping the tombstone inside the buffer.

If RAII guard works for most of our use cases, I think it makes sense to support it. Maybe we can even get rid of shared channels as I quickly skimmed through the code and I don't see a place where we need to re-use a channel that was opened previously somewhere else (I could've missed something).

**Comment by thurstonsand** [comment]
I definitely prefer RAII guard if implementation is possible. users don't even have to really think about it (assuming they don't ever interact with mutable global state)

**Comment by martinthomson** [comment]
For those channels that are ordered (which I believe is the case between the infrastructure components and network), closing seems to be the most elegant solution.  If that means a per-channel sink, that might be more work, but I have no objections (Alex and Thurston might though).

Runtime checks are acceptable here.  I don't quite know how we can do anything reasonable at compile time here.  There might be, but unless you have a really elegant solution, I wouldn't try to start there.

**Comment by thurstonsand** [comment]
the `HttpNetwork` won't always keep a running count of record id. right now, that's necessary due to the way these are batched in the `Http` layer, but this will eventually move to a `Stream`ing model, where ordering is guaranteed, and so counting records is no longer needed. Once that happens, the network layer will not have any knowledge of record id.

**Comment by akoshelev** [comment]
>  If that means a per-channel sink, that might be more work, but I have no objections (Alex and Thurston might though).

Per-channel sink is what we want. It was a mistake to start with single channel between infra and network, so I am 100% supportive of making this change

**Comment by andyleiserson** [comment]
> Per-channel sink is what we want.

For the outbound direction only, or for both inbound and outbound?

If the concern is enabling reordering of data for different channels, it seems like the receive buffer structure already provides that. Is the idea that the receive buffer structure might change in the future to something that has limited capacity per channel? If yes, then I would say we want sink-per-channel for both inbound and outbound handoff between gateway and network.

**Comment by martinthomson** [comment]
This probably applies in both directions eventually as that provides a more direct coupling of consumption and production for back pressure purposes.  Any time we add a buffering step or redistribution stage, we take on an overhead that might be avoided.

Happy to take changes in stages.

**Comment by andyleiserson** [comment]
I have some WIP using an interface like:

```
pub trait SinkProvider {
    type Sink: futures::Sink<Vec<u8>, Error = Error> + Send + Unpin + 'static;
    fn get_sink(&self, channel_id: &Channel) -> Self::Sink;
}
```

... for both the inbound and outbound directions. For the outbound direction, gateway would call this method on network (or some proxy object provided by the network). For the inbound direction, network would call the method on gateway (or proxy). In both cases, the idea is that calling `get_sink` is an "open channel" operation. The sender stores the sink somewhere for the duration that the channel is needed, and closes the sink when it has sent all records.

High level, my rationale for doing it this way is that pushing data in this way is likely to be more efficient than pulling data, and reduces the number of places that we need to spawn tasks that move messages around. To start with, I wouldn't change the structure of the gateway implementation significantly, so inbound messages would arrive sink-per-channel from network, over a single channel in gateway, to the receive buffer, and outbound messages would go from the send buffer, over a single channel in gateway, to the network over sink-per-channel. 

Thoughts on this approach?

Another approach I considered was putting some sort of Stream/Sink-like interface on the SendBuffer and ReceiveBuffer so that the network implementation could directly push/pull data. But that seemed like a larger change and the SendBuffer and ReceiveBuffer are currently not async at all (the async logic is all in the Gateway where they are instantiated).

**Comment by akoshelev** [comment]
I like it too. Wondering if #352 has an impact on this design? it is likely that `Network` will no longer be a trait but rather just a tiny proxy layer on top of `Transport`

**Comment by thurstonsand** [comment]
also, network will no longer have a `Sink`, but will just have a `send` fn exposed

**Comment by andyleiserson** [comment]
Yes, I do think this is closely related to #352. How do you envision per-channel sinks working with the `fn send` in #352?

**Comment by akoshelev** [comment]
we could add the `ChannelId` argument to the `send` function. It does not seem to be a problem to expose a sink from it too,  `Network` struct would just need to multiplex data to a single `Transport::send` channel. I would imagine `Transport` to eventually expose per-channel streams

**Comment by akoshelev** [comment]
Ok, @andyleiserson and I have discussed this issue in context of sharding. Our current sharding implementation uses the same infrastructure as MPC channels, but opens up a lot more of them (one per destination shard). Shard channels tend to require more memory because shards often stream the whole report between them, making this issue even more prevalent on MPC executions that involve large number of shards.

**Comment by akoshelev** [comment]
There was an attempt to implement receiver closure, but there were some subtle issues with it https://github.com/private-attribution/ipa/pull/805
